### DISCUSSION TITLE
Layer 2 multicast in INET

### DISCUSSION THREAD

**Discussion by paintspot87** [omnetpp/omnetpp#1107] [Q&A] [answered]
Hi,

What would be a quick and dirty way to implement a multicast functionality in Omnet INET for layer 2 traffic? Multicast is available for UDP apps but I can't see that it is available for layer 2 apps.

Currently, I have modified the StandardHost.ned and connected an application directly to the LLC but I want to be able to multicast the traffic to specific hosts rather than unicast it. 

Any ideas?

**Comment by rhornig** [comment, marked as answer]
INET does support layer 2 multicast. There is an `addMulticastMacAddress()` method on the `NetworkInterface` class. Once you add a mac address to an interface it handles all incoming packets correctly and passes up any packets addressed to the multicast address. 

There is even an `Ethernet(Source|Sink)App.io.localAddress` parameter that adds the specified mac address as a multicast address to the network interface so you can specify the mac address per application (if you happen to install more than one app in a host). 

You don't have to modify the StandardHost either. If you install an Ethernet(Source|Sink)App inside a   StandardHost, it will use a layer 2 socket and the app layer packets will be sent directly to layer 2  routing around the transport and network layers.

**Reply by paintspot87** [reply]
Thank you so much for the quick reply. I will try this out today.

**Comment by paintspot87** [comment]
Hi again, I'm running into some issues while using the EthernetSource and Sink applications.

In the **receiving node** I get this error:

`<!> handlePacket(): Unknown protocol: protocolId = 89, protocolName = unknown, servicePrimitive = INDICATION, pathStartGate = MultiCast.MU.ethernet.macHeaderChecker.out, pathEndGate = MultiCast.MU.bl.in[2] -- in module (inet::MessageDispatcher) MultiCast.MU.bl (id=286), at t=0.0000177s, event #20`

Should the source and sink apps be connected to an LLC-socket or what am i doing wrong?


Also, when I try to specify an io.localAddress in any node I get this error:

`<!> check_and_cast(): Cannot cast (inet::Request*)BIND to type 'inet::Packet *' -- in module (inet::EthernetMacHeaderInserter) MultiCast.IDE_A.ethernet.macHeaderInserter (id=61), at t=0s, event #10`

I would be grateful for any tips on what to focus on as I am stumbling a bit in the dark right now :).

These are my .ned and .ini-files

```
import inet.networks.base.WiredNetworkBase;
import inet.node.ethernet.Eth100M;
import inet.node.inet.StandardHost;
import inet.node.ethernet.EthernetSwitch;

network MultiCast extends WiredNetworkBase
{
    submodules:
        IDE_A: StandardHost {
            parameters:
                @display("p=298,77");
        }
        IDE_B: StandardHost {
            parameters:
                @display("p=391,77");
        }
        IDE_C: StandardHost {
            parameters:
                @display("p=501,77");
        }
        MU: StandardHost {
            parameters:
                @display("p=391,272");
        }
        switch: EthernetSwitch {
            parameters:
                @display("p=391,169");
        }
    connections:
        IDE_A.ethg++ <--> Eth100M <--> switch.ethg++;
        IDE_B.ethg++ <--> Eth100M <--> switch.ethg++;
        IDE_C.ethg++ <--> Eth100M <--> switch.ethg++;
        MU.ethg++ <--> Eth100M <--> switch.ethg++;
}
```

```
[General]
sim-time-limit = 10s
*.*.eth[*].bitrate = 100Mbps
*.*.ethernet.typename = "EthernetLayer"
*.*.eth[*].typename = "LayeredEthernetInterface"

[Config MultiCast]
network = MultiCast
**.IDE_A.numApps = 1
**.IDE_A.app[0].typename = "EthernetSourceApp"
**.IDE_A.app[0].source.packetLength = 84B 
**.IDE_A.app[0].source.productionInterval = 1s
#**.IDE_A.app[0].io.localAddress = "01:0C:CD:01:00:01"
**.IDE_A.app[0].io.remoteAddress = "MU"
**.IDE_A.app[0].io.interface = "eth0"


**.IDE_B.numApps = 1
**.IDE_B.app[0].typename = "EthernetSinkApp"
**.IDE_B.app[0].io.interface = "eth0"

**.IDE_C.numApps = 1
**.IDE_C.app[0].typename = "EthernetSinkApp"
**.IDE_C.app[0].io.interface = "eth0"

**.MU.numApps = 1
**.MU.app[0].typename = "EthernetSinkApp"
**.MU.app[0].io.interface = "eth0"
```

**Reply by levy** [reply]
There's no error for me in INET master. Which version are you using?

**Reply by paintspot87** [reply]
Omnet 6.0.2 and INET 4.5.2 on Windows 10.

I will try this on a Ubuntu VM too.

**Comment by paintspot87** [comment]
So I understand now that my general settings messed up and that the formatting isn't as it shows here on the forum. 

`*.*.ethernet.typename = "EthernetLayer"`
`*.*.eth[*].typename = "LayeredEthernetInterface"`

Produced the errors.

Now I try to define localAddresses to all nodes and remoteAddress from my source but it becomes broadcasted to all nodes instead.. Why is that?


[omnetpp.txt](https://github.com/omnetpp/omnetpp/files/14373277/omnetpp.txt)

**Reply by levy** [reply]
That's because the MAC forwarding tables don't contain any rules for the multicast addresses.

**Reply by paintspot87** [reply]
Thank you, I solved it by doing the following:

You have saved me a lot of headache with my thesis, much appreciated.

```
**.hasGlobalArp = false
**.IDE_A.app[0].io.remoteAddress = "01:00:00:00:00:03"
**.switch.macTable.forwardingTable = [{address: "01:00:00:00:00:03", interface: "eth1"}, \
                                      {address: "01:00:00:00:00:03", interface: "eth2"}, \
                                      {address: "01:00:00:00:00:04", interface: "eth2"}, \
				      {address: "01:00:00:00:00:05", interface: "eth3"}]
```
